ports/stm32/boards: Add NUCLEO-L152RE board support.

yn386 opened the pull request:
### Detail of changes:
This change adds NUCLEO-L152RE support to the STM32 port.
NUCLEO-L152RE: https://www.st.com/en/evaluation-tools/nucleo-l152re.html
This board use STM32L152RE: https://www.st.com/en/microcontrollers-microprocessors/stm32l152re.html.

STM32L1 port is discussed before here: https://forum.micropython.org/viewtopic.php?f=12&t=5914&p=33820
but I was not be able to find any STM32L1 implementation by searching.
So I tried to implement STM32L1 port with reference to another NUCLEO board such as F4 and L4.

I changed submodulelib/stm32ilb to my fork temporary, but I posted a pull request
https://github.com/micropython/stm32lib/pull/21 .
If this pull request is merged, I will change .gitmodule.

### Tested:
REPL on USART2 works.
Run run-tests.py and it works as expected.

I checked some part of peripherals:
GPIO/Timer/I2C/SPI/ADC/DAC
but not tested all peripherals.

### Limitation:
ADCAll.read_core_vbat() throws NotImplementedError exception because STM32L152RE does not have VBAT.
16KB EEPROM is not used.

yn386 commented:
I noticed writing file into internal flash file system is too slow.
I fixed it.

yn386 commented:
I found why pyb.i2c() with dma=True does not work.
I fixed it and removed known problem.

dpgeorge commented on ports/stm32/modmachine.c:
```diff
@@ -351,7 +351,14 @@ MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(machine_freq_obj, 0, 4, machine_freq);
 STATIC mp_obj_t machine_idle(void) {
+    #if defined(STM32L1)
+    mp_uint_t irq_state;
+    irq_state = MICROPY_BEGIN_ATOMIC_SECTION();
```
why did you need to take the atomic state here?

yn386 replied:
For STM32L1, IRQ state may be changed after __WFI() only in MCU debug component (DBGMCU) is enabled.
This causes unexpected HardFault. To prevent it, I needed to save and restore IRQ state beetween __WFI().

Usually DBGMCU is enabled only when debugging with OpenOCD or using STM32CubeProgrammer with enableing 'Debug in Low Power mode'.

dpgeorge replied:
I'll have to understand more about this.  Is it only for the STM32L1xx series?  Is there an application note (or some documentation) about why and when it's needed?

I don't want to add this for all ports, especially in the `MICROPY_EVENT_POLL_HOOK` macro.

yn386 replied:
> Is it only for the STM32L1xx series?

Maybe, yes. I didn't meet such HardFault with NUCLEO-F446RE.
So I added these implementation for STM32L1 only.

> Is there an application note (or some documentation) about why and when it's needed?

I'm sorry that I couldn't find any documentation, but there are some similar issues.
For STM32L1xx, program counter seems to corrupt after wfi when irq and DBGMCU are enabled.
https://community.st.com/s/question/0D50X00009XkbFuSAJ/wfi-in-run-mode-causing-hard-fault
https://github.com/probe-rs/probe-rs/issues/350
https://github.com/RIOT-OS/RIOT/issues/11820
https://github.com/RIOT-OS/RIOT/pull/11830

> I don't want to add this for all ports, especially in the MICROPY_EVENT_POLL_HOOK macro.

This issue is due to enabling DBGMCU, so disabling DBGMCU instead of taking atomic state can also prevent this issue.
To disable it, execute
`DBGMCU->CR &= ~(DBGMCU_CR_DBG_SLEEP | DBGMCU_CR_DBG_STOP | DBGMCU_CR_DBG_STANDBY);`
during initialization of STM32L1.
Even if disabling DBGMCU, I could debug by using ST-LINK or OpenOCD.
Should I fix it?

dpgeorge replied:
To be able to move forward with this PR (merge it) I suggest disabling DBGMCU during initialisation of the STM32L1xx, and not changing the WFI behaviour of other ports.

That will at least keep things the same for existing MCUs, and get it working on STM32L152.

Then we can possible fix/improve things in the future.

yn386 replied:
Thank you for reply. I added some commits.

- I removed taking atomic state between __WFI() and add disabling DBGMCU at initialization (49c66271).
- When using STM32CubeProgrammer to upload firmware, the user must disable 'Debug in Low Power Mode' (dLPM).
Also, STM32CubeProgrammer often cannot detect NUCLEO-L152RE, so adding mode 'Under Reset' (mode=UR).
To explain it, I added deploy.md for NUCLEO-L152RE (8d520538, d248e3b7).
- While writing deploy.md, I also noticed that machine.bootloader() was not working.
Now I fixed it, so please include this commit (e14f2627) to this PR.
Sorry for inconvenience.

dpgeorge commented on ports/stm32/mpconfigport.h:
```diff
@@ -271,12 +271,24 @@ static inline mp_uint_t disable_irq(void) {
+        mp_uint_t irq_state; \
         extern void mp_handle_pending(bool); \
         mp_handle_pending(true); \
+        irq_state = MICROPY_BEGIN_ATOMIC_SECTION(); \
```
why does it take the atomic state?

yn386 replied:
This is the same reason with modmachine.c.

dpgeorge commented:
Thank you for this contribution.  I have tested it on the NUCLEO_L152RE board and it works.

To get this into a mergable state, can you please split up the main commit (that adds L1 support) into 3 separate commits:
1. update the stm32lib submodule
2. modify the code in `ports/stm32` to support L1 MCUs
3. add the `ports/stm32/boards/NUCLEO_L152RE` board definition

yn386 committed:
lib/stm32: Add STM32L1xx platform.

yn386 committed:
ports/stm32: Add STM32L1 support.

This change adds STM32L1 support to the STM32 port.
STM32L152RE: https://www.st.com/en/microcontrollers-microprocessors/stm32l152re.html

yn386 committed:
ports/stm32/boards: Add NUCLEO-L152RE board support.

This change adds NUCLEO-L152RE support to the STM32 port.
NUCLEO-L152RE: https://www.st.com/en/evaluation-tools/nucleo-l152re.html
This board use STM32L152RE: https://www.st.com/en/microcontrollers-microprocessors/stm32l152re.html

yn386 commented:
Thank you. I rebased and split into 3 separate commits.

dpgeorge commented:
> I changed submodulelib/stm32ilb to my fork temporary, but I posted a pull request
> [micropython/stm32lib#21](https://github.com/micropython/stm32lib/pull/21) .
> If this pull request is merged, I will change .gitmodule.

The L1 support has been merged in stm32lib, on branch `work-F0-1.9.0+F4-1.16.0+F7-1.7.0+G0-1.5.1+G4-1.3.0+H7-1.6.0+L0-1.11.2+L1-1.10.3+L4-1.17.0+WB-1.10.0+WL-1.1.0`, so you can update the git submodule in this PR.

Also, please add a `MICROPY_PLATFORM_VERSION` for L1 in `ports/stm32/mphalport.h`.

yn386 commented:
Thank you. I rebased and
- Updated `.gitmodule `and `lib/stm32` (f9f479fb).
- Added `MICROPY_PLATFORM_VERSION` for L1 in ports/stm32/mphalport.h (e3fb8d48).
https://github.com/micropython/micropython/blob/e3fb8d48f046e6f2fc934699dc12ad8c5ffdc65c/ports/stm32/mphalport.h#L20-L21

dpgeorge commented on ports/stm32/boards/stm32l152xe.ld:
```diff
@@ -0,0 +1,39 @@
+    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 512K /* entire flash */
+    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 16K  /* sector 0-3 */
+    FLASH_FS (rx)   : ORIGIN = 0x08004000, LENGTH = 112K /* sectors 4-31 */
+    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 384K /* sectors 32-127 */
```
It looks like the L152 has uniform 512-byte erase page size (is that correct?).  So it would be simpler to put the filesystem at the end of flash, eg:
```
    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 512K /* entire flash */
    FLASH_FS (rx)   : ORIGIN = 0x08064000, LENGTH = 112K /* sectors 912-1023 */
```
Then the board needs to specify `boards/common_basic.ld` in `mpconfigboard.mk`, and no need to specify `TEXT1_ADDR` (because the firmware is contiguous in flash, starting at `TEXT1_ADDR`).

yn386 replied:
> It looks like the L152 has uniform 512-byte erase page size (is that correct?).

No, L152 has uniform 256-byte page size.
Also, L152 has uniform 4KB sectors (16 pages per sector).

> So it would be simpler to put the filesystem at the end of flash

I understand. I fixed it.

dpgeorge commented:
Thanks for updating the filesystem location.

Rebased and merged in ae0b0e701899297d057ab8dc578b3bdbf0144fc4 through e6d351318746495bf88d2c4bd7cbd81e94a2290e, with some very minor changes to code formatting.

yn386 commented:
Thank you very much for reviewing. I'm glad this PR was merged.
